### ISSUE TITLE
Support CRUD actions to records in shared zones

### ISSUE THREAD

**Issue by mdb** [vinyldns/terraform-provider-vinyldns#55]
To make updates to records that live in a zone that is `shared=true` (See [VinylDNS docs](https://www.vinyldns.io/api/zone-model#shared-zones)), it is necessary to use the [batch change API endpoint](https://www.vinyldns.io/api/create-batchchange), which `terraform-provider-vinyldns` does not currently do. 

Rather than shoehorn support for recordsets that live in shared zones into the current `vinyldns_record_set` model, I think this warrants the creation of a `vinyldns_batch_changes` resource:
https://www.vinyldns.io/api/batchchange-model.html

**Comment by mdb** [comment]
If the batch change response for a shared zone returns a record ID that the TF provider is authorized to do a lookup on to get up-to-date/current info about the record set, then I believe use of the batch changes API could work in performing CRUD actions on recordsets in `shared=true` zones. If this is not feasible, then use of the batch API may not be possible, at least not without some acrobatics.  

@pauljamescleary @rebstar6 What is the the feasibility of looking up current details about a record set by ID if it lives in a `shared=true` zone?

**Comment by pauljamescleary** [comment]
@mdb - if you are the owner of a record (you created or updated it in a batch change and a member of the record owner group) then you CAN do a `getRecordSet` and see it

**Comment by mdb** [comment]
@pauljamescleary Great - I assume you are referring to `ownerGroupId` on [batch change summary](https://www.vinyldns.io/api/batchchange-model.html#batch-change-attributes-). Am I correct in thinking that one can...

* use the batch changes API to create or update records in a `shared=false` zone if the user is part of the zone's admin group
* use the batch changes API to create or update records in a `shared=true` zone if the recordset has now `ownerGroupId` & the user specifies an `ownerGroupId` for a group in which they have admin membership on the request?

?

**Comment by pauljamescleary** [comment]
@mdb yes.  In addition to "user is part of zone's admin group", we have zone acl's, so a user maybe granted access to a *private* zone via ACL (i.e. the user does not HAVE to be a member of the zone admin group, they can access a private zone in another way).

When the record set is saved in our database, we attach the `ownerGroupId` to the record set itself.  When you do a `getRecordSet` one of the access checks we make is "if the zone is shared and the user is a member of the owner group on the record set, then let them get it."

As an aside, I still think we should have a simpler API too.  The record FQDN + record type forms a unique key.  The original API design did not account for this (mostly because I didn't know DNS that well at the time).

We should be able to say:

`PUT /recordsets/<fqdn>/<recordtype>` and have it do an idempotent put.

Similarly we should be able to

`GET /recordsets/<fqdn>/<recordtype>` and anyone should be able to query it.

**Comment by rhenning** [comment]
Thanks for clarifying this @pauljamescleary.

At quick glance, it seems like a [batch change response](https://www.vinyldns.io/api/batchchange-model.html#batch-change-example-) includes the necessary IDs required to do a `[getRecordSet](https://www.vinyldns.io/api/get-recordset.html)` so I believe that we could conceivably use the batch change API for all recordSet actions, shared or not.

**Comment by pauljamescleary** [comment]
@rhenning the main issue with using Batch Change is that we are introducing a "Manual Review" state.  This is needed primarily for zone adds (an FQDN or IP address zone does not exist).  

I think that is a complication of using batch changes in TF.  The turnaround for things like "Manual Review" is 24 hours.

**Comment by mdb** [comment]
@pauljamescleary Will "Manual Review" be an _option_ or something that happens for all "zone adds (an FQDN or IP address zone does not exist)" scenarios?

**Comment by britneywright** [comment]
Manual Review is an option, there is an "allowManualReview" parameter on the create batch change endpoint that defaults to true https://www.vinyldns.io/api/create-batchchange.html. If it is set to false and there are any "permissible" errors (errors that would normally result in the batch change going into Manual Review) the batch change will be returned with those errors. If the batch change doesn't have any errors it will be processed.
